Add bluetooth host buf unit test

ahmedmoheb-nordic opened the pull request:
A set of unit test projects for 'bluetooth/host/buf.c'.
These unit test projects should
1 - Test the behavior of module APIs
2 - Increase the test coverage report percentage of the module

thoh-ot commented on tests/bluetooth/host/buf/bt_buf_get_evt/kconfig_override.h:
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2021 Demant
```
```suggestion
 * Copyright (c) 2022 Nordic Semiconductor ASA
```

ahmedmoheb-nordic replied:
Hi @thoh-ot , Could you review again? That change request has been applied long time ago.

ahmedmoheb-nordic replied:
@thoh-ot Could you review again? Your change request is blocking the PR

alwa-nordic commented:
RFC on something similar to property based testing: We should have a collection of statements that are always true, and parameterize them over all the variables that may affect the UUT.

Here are some example properties for `bt_buf_get_cmd_complete`:

- Any calls to `net_buf_alloc_fixed` pass on `timeout`.
- If `net_buf_alloc_fixed` returns non-null, that buffer is returned.
- If `bt_dev.sent_cmd == NULL`, `net_buf_alloc_fixed` is used.
- The returned value must not be found in `bt_dev.sent_cmd` afterwards unless it's NULL.
- If `bt_dev.sent_cmd != NULL`, that value is returned.
- If `CONFIG_BT_HCI_ACL_FLOW_CONTROL` is set, any call to `net_buf_alloc_fixed` uses `bt_buf_get_evt_pool()`.
- If `CONFIG_BT_HCI_ACL_FLOW_CONTROL` is not set, any call to `bt_buf_get_hci_rx_pool` uses `bt_buf_get_evt_pool()`.

We express these in parameterized functions and select interesting combinations of values for those parameters, like sketched below.

```c
/* This test should pass for for all values of bt_dev__sent_cmd, timeout, net_buf_alloc_fixed_return */
void test_bt_dev_sent_cmd_cleared_parameterized(bt_dev__sent_cmd, timeout, net_buf_alloc_fixed__retval) {
  bt_dev.sent_cmd = bt_dev__sent_cmd;
  ztest_returns_value(net_buf_alloc_fixed, net_buf_alloc_fixed__retval);
  struct net_buf * returned = bt_buf_get_cmd_complete(timeout);
  if (bt_dev.sent_cmd != NULL) {
    zassert_not_equal(bt_dev.sent_cmd, returned, "The returned buffer is still stored in `sent_cmd`");
  }
}
void test_bt_dev_sent_cmd_cleared() {
  /* Test with interesting values of the "forall" variables. */
  /* Here we use loops to create a test matrix, but hand-picked combinations may be just as good. */
  for (bt_dev__sent_cmd <- NULL, 'a')
  for (net_buf_alloc_fixed__retval <- NULL, 'b')
  for (timeout <- 0, 10) {
    test_sent_cmd_cleared_parameterized(bt_dev__sent_cmd, timeout, net_buf_alloc_fixed__retval);
  }
}
```

alwa-nordic reviewed (changes requested):
Please put the copying of the mocks in it's own commit. Otherwise it's hard for us to see what has changed.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,145 @@
+	ztest_expect_value(net_buf_alloc_fixed, pool, bt_buf_get_hci_rx_pool());
+#endif
+
+	ztest_expect_value(net_buf_simple_reserve, buf, &test_reference.b);
```
In my opinion, if we are going to override an inline function, I'll vote for editing the current header file to allow overriding.

Creating a new header file will require copy the original file content and modify the declaration of the inline functions only.
That will create some sort of inconsistency and it will be harder to maintain the copies compatible over time.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,145 @@
+ *     - net_buf_alloc_fixed() returns a NULL value
+ *
+ *  Expected behaviour:
+ *     - net_buf_alloc_fixed() to be called with the correct memory allocation pool
+ *       and the same timeout value passed to bt_buf_get_cmd_complete()
+ *     - bt_buf_get_cmd_complete() returns NULL
```
I think that can be required if the API under testing accepts single parameter which the timeout value.
Splitting it into more than test function will create some redundancy as we won't expect different outputs for the same inputs.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,145 @@
+
+	bt_dev.sent_cmd = NULL;
+
+#if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
```
By splitting it up and using the configuration parameter as a precondition, we will introduce again the #if checking again, correct?

alwa-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,145 @@
+
+	bt_dev.sent_cmd = NULL;
+
+#if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
+	ztest_expect_value(net_buf_alloc_fixed, pool, bt_buf_get_evt_pool());
+#else
+	ztest_expect_value(net_buf_alloc_fixed, pool, bt_buf_get_hci_rx_pool());
+#endif
```
We should try to make the branching (including ifdefs) as simple as possible in test. Here we can define a variable `const expected_pool` that get's its value set depending on the kconfig.

```suggestion
#if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
	const struct net_buf_pool *expected_pool = bt_buf_get_evt_pool();
#else
	const struct net_buf_pool *expected_pool = bt_buf_get_evt_pool();
#endif
ztest_expect_value(net_buf_alloc_fixed, pool, expected_pool);
```

jori-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,145 @@
+#endif
+
+	ztest_returns_value(net_buf_alloc_fixed, NULL);
+	ztest_expect_value(hooks_net_buf_alloc_fixed_timeout_validation_hook, value, timeout.ticks);
```
@ahmedmoheb-nordic see e.g. #48074 for an example.

jori-nordic replied:
If you don't want to use FFF, I'd suggest renaming that fn to `net_buf_alloc_timeout_value`, since the function is just a mock to extract the tick value and you do the mock setup in the `main.c`.

jori-nordic commented on tests/bluetooth/host/mock_host/src/iso.c:
```diff
@@ -0,0 +1,43 @@
+#include <zephyr/bluetooth/iso.h>
+#include <host/hci_core.h>
+
+
```
can you add a comment here ? 
```c
/* This file doesn't contain mocks, but rather fakes of the necessary parts of
 * subsys/bluetooth/host/iso.c. The API implementations that are copied here
 * should be kept in sync with the original.
 */
```

jori-nordic replied:
also see https://www.martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs
and
http://xunitpatterns.com/Test%20Double.html

ahmedmoheb-nordic replied:
Thanks for sharing these pages, they were really helpful

jori-nordic commented on tests/bluetooth/host/mock_host/src/net.c:
```diff
@@ -0,0 +1,53 @@
+/*
```
can you rename that file to `buf.c` ?

ahmedmoheb-nordic replied:
These definitions exist in subsys/net/buf.c.
I moved it to net/buf.c.

jori-nordic commented on tests/bluetooth/host/mock_host/src/net.c:
```diff
@@ -0,0 +1,53 @@
+
+struct net_buf *net_buf_alloc_fixed(struct net_buf_pool *pool, k_timeout_t timeout)
+{
+	zassert_not_null(pool, "Value was NULL");
```
redundant message. Changing it to "Invalid pool" would be better (or even removing it is fine).

jori-nordic replied:
I also think this check might be un-necessary, seeing how you're checking its value anyways in the next line. So the test case function should take care of checking if `pool` is `NULL` or not IMO.

ahmedmoheb-nordic replied:
Correct, It's a redundant check.
I think it was used as a test in the beginning but it was left over.

jori-nordic commented on tests/bluetooth/host/mock_host/src/net.c:
```diff
@@ -0,0 +1,53 @@
+void net_buf_simple_reserve(struct net_buf_simple *buf, size_t reserve)
+{
+	ztest_check_expected_value(buf);
+	zassume_equal(reserve, BT_BUF_RESERVE, "Invalid value");
```
why is that assumption here ? I think this should rather be a `ztest_check_expected_value(reserve)`. If it's a default value you want to check for, you should use an assert with `BT_BUF_RESERVE`. 

But I think you should setup the mocking properly for that param too in the test cases instead of hardcoding an assert.

The need for suite-based defaults might be solved by FFF (haven't read the docs yet). I know CMocka has a semi-documented feature for that (using negative value as counts, see [will_return_count](https://api.cmocka.org/group__cmocka__mock.html#gaac9a382a8509489d619538fe264e66a7), so maybe FFF also does.

When that is supported you can just have a suite setup fn that sets up those default mocks and then not worry about it in the individual testcases, while not having to hardcode values in the mocks themselves.

jori-nordic reviewed (changes requested):
A few comments. 
Could you also have a look at include statements in all files, and remove anything that's not actively used in the file ?

yperess reviewed (changes requested):
Please do not use ztest_mock, it's actively being removed. Use FFF instead.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,170 @@
+	zassert_equal(type, BT_BUF_EVT,
+		      "bt_buf_get_cmd_complete() returned incorrect buffer type %u, expected %u (%s)",
+		      bt_buf_get_type(buf), BT_BUF_EVT, STRINGIFY(BT_BUF_EVT));
+	zassert_equal(bt_buf_get_type(buf), type,
```
This inline function is so simple that it can be checked while testing others APIs.
Creating a dedicated project for it won't add a value because it will require calling other functions (e.g. bt_buf_get_rx()) to set the type and then check that it was set correctly which is the same logic here.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,170 @@
+ *  Expected behaviour:
+ *     - net_buf_alloc_fixed() isn't called
+ *     - bt_dev.sent_cmd value isn't changed after calling bt_buf_get_cmd_complete()
+ *     - bt_buf_get_cmd_complete() returns the same value returned by net_buf_alloc_fixed()
```
No, so impressed that you caught it

ahmedmoheb-nordic replied:
No

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/src/main.c:
```diff
@@ -0,0 +1,170 @@
+	zassert_equal(bt_buf_get_type(buf), type,
+		      "bt_buf_get_type() returned incorrect type %u, expected %u",
+		      bt_buf_get_type(buf), type);
+	zassert_true((bt_dev.sent_cmd == before_call_sent_cmd_value),
```
which comment, could you clarify?

ahmedmoheb-nordic commented on tests/bluetooth/host/mock_host/src/hci_core.c:
```diff
@@ -0,0 +1,38 @@
+
+static void init_work(struct k_work *work);
+
+struct bt_dev bt_dev = {
```
One of this struct fields is used by 'bt_buf_get_cmd_complete()'.
Undefined reference error is generated if that definition isn't there.

jori-nordic commented:
> Please do not use ztest_mock, it's actively being removed. Use FFF instead.

Yes we know about it. As @ahmedmoheb-nordic is still ramping up on zephyr, we'll try to merge it with ztest (as the PR is almost in a mergeable state), to get something running in CI, and the very next PR (next week or so) will be migrating this test to FFF.
Is this okay with you ?

alwa-nordic commented on tests/bluetooth/host/mock_host/src/hci_core.c:
```diff
@@ -0,0 +1,28 @@
+
+static void init_work(struct k_work *work);
+
+struct bt_dev bt_dev = {
+	.init          = Z_WORK_INITIALIZER(init_work),
+#if defined(CONFIG_BT_DEVICE_APPEARANCE_DYNAMIC)
+	.appearance = CONFIG_BT_DEVICE_APPEARANCE,
+#endif
+};
```
@jori-nordic meant that you don't need the initializer.

```suggestion
/* Mock */
struct bt_dev bt_dev;
```

alwa-nordic commented on tests/bluetooth/host/buf/bt_buf_get_type/src/main.c:
```diff
@@ -0,0 +1,96 @@
+ *  LUT containing testing parameters that will be used
+ *	during each iteration to cover different scenarios
+ */
+static const enum bt_buf_type testing_params_lut[TEST_PARAMETERS_LUT_ROWS_COUNT] = {
+	/** HCI command */
+	BT_BUF_CMD,
+	/** HCI event */
+	BT_BUF_EVT,
+	/** Outgoing ACL data */
+	BT_BUF_ACL_OUT,
+	/** Incoming ACL data */
+	BT_BUF_ACL_IN,
+	/** Outgoing ISO data */
+	BT_BUF_ISO_OUT,
+	/** Incoming ISO data */
+	BT_BUF_ISO_IN,
+	/** H:4 data */
+	BT_BUF_H4,
+};
```
If you want to verify the size of the array, it's not enough to specify its length. You will not get a compiler warning if the specified length is larger than the initializer.

```suggestion
static const enum bt_buf_type testing_params_lut[] = {
	/** HCI command */
	BT_BUF_CMD,
	/** HCI event */
	BT_BUF_EVT,
	/** Outgoing ACL data */
	BT_BUF_ACL_OUT,
	/** Incoming ACL data */
	BT_BUF_ACL_IN,
	/** Outgoing ISO data */
	BT_BUF_ISO_OUT,
	/** Incoming ISO data */
	BT_BUF_ISO_IN,
	/** H:4 data */
	BT_BUF_H4,
};
BUILD_ASSERT(ARRAY_SIZE(testing_params_lut) == TEST_PARAMETERS_LUT_ROWS_COUNT);
```

ahmedmoheb-nordic replied:
Actually, the array size was specified to be used with 'LISTIFY' asit doesn't accept the 'ARRAY_SIZE' as a parameter.

On the other hand, trying to exceed the specified size will result in 'excess elements in array initializer' warning.

But your suggestion is a good one, I didn't know about it.

jori-nordic replied:
yep I thought it'd work but the definition of ARRAY_SIZE is C code, and not a C macro, so the preprocessor doesn't actually know the size of the array.

alwa-nordic replied:
(Joking) It can always be solved with [more macros].

[more macros]: https://www.boost.org/doc/libs/1_39_0/libs/preprocessor/doc/ref/list_size.html

jori-nordic commented on tests/bluetooth/host/mock_host/src/net/buf.c:
```diff
@@ -0,0 +1,48 @@
+struct net_buf *net_buf_alloc_fixed(struct net_buf_pool *pool, k_timeout_t timeout)
+{
+	ztest_check_expected_value(pool);
+	utils_net_buf_alloc_timeout_value_mock(timeout.ticks);
```
thinking a bit more about it, my opinion is that this function should be defined here, since it's essentially a mock: It only exists to workaround a limitation of the testing framework, but it is checking parameter values of `net_buf_alloc_fixed`. 

You can keep the declaration in `buf_help_utils.h` though.

jori-nordic reviewed (changes requested):
Some more comments. Once those are addressed, we will rewrite the git history to be a bit cleaner and we should be ready to merge.

ahmedmoheb-nordic commented on tests/bluetooth/host/mock_host/src/assert.c:
```diff
@@ -0,0 +1,37 @@
+	if (mock_check_if_assert_expected() == true) {
+		printk("Assertion expected as part of a test case.\n");
+		/* This will mark the test as passed and stop execution */
+		ztest_test_pass();
```
I agree with you that the fail scenario works and if there is an unexpected assert, calling `mock_check_if_assert_expected()` will terminate the test and mark it as failed.

The problem I encountered while testing the invalid parameters.
Simply the execution continues after the the failed assertion which will hit parts of the code that the test isn't handling.

For example, it will try to allocate memory through memory pools and the test doesn't set these parameters correctly.

jori-nordic replied:
ah yeah then I understand the problem. You can leave as-is then. Just add one more line to the comment to make it clear:
```c
/* This will mark the test as passed and stop execution:
 * Needed in the passing scenario to prevent undefined behavior after hitting the assert.
 * In real builds (non-UT), the system will be halted by the assert.
 */

alwa-nordic commented on tests/bluetooth/host/common/defaults_cmake.txt:
```diff
@@ -0,0 +1,55 @@
+  add_definitions(-include ${KCONFIG_OVERRIDE_FILE})
+endif()
+
+add_definitions(-include ztest.h)
+add_definitions(-include soc.h)
```
Please instead add `#include`s to each file that needs this.

ahmedmoheb-nordic replied:
It seems that some other files (source and header files) that don't exist in the project directory depends on the first one `add_definitions(-include ztest.h)` and causes the compilation process fails.

I removed the second one as it doesn't have any effect

alwa-nordic commented on tests/bluetooth/host/common/defaults_cmake.txt:
```diff
@@ -0,0 +1,55 @@
+  ${ZEPHYR_BASE}/subsys/testsuite/ztest/include/zephyr
+)
+
+FILE(GLOB host_sources
+  ${ZEPHYR_BASE}/subsys/bluetooth/host/buf.c
+)
```
This does not look global to `test/bluetooth/host`. Please move this to folder `buf`, or to each test.

alwa-nordic commented on tests/bluetooth/host/common/defaults_cmake.txt:
```diff
@@ -0,0 +1,55 @@
+  ${ZEPHYR_BASE}/subsys/bluetooth/host/buf.c
+)
+
+FILE(GLOB mock_sources
+  ${ZEPHYR_BASE}/tests/bluetooth/host/mock_host/src/net/buf.c
+  ${ZEPHYR_BASE}/tests/bluetooth/host/mock_host/src/iso.c
+  ${ZEPHYR_BASE}/tests/bluetooth/host/mock_host/src/hci_core.c
+  ${ZEPHYR_BASE}/tests/bluetooth/host/mock_host/src/assert.c
+)
```
This is also not common to `test/bluetooth/host`. (If we want to test `hci_core.c`, we don't want to mock it.)

jori-nordic commented on tests/bluetooth/host/buf/mocks/kconfig.h:
```diff
@@ -0,0 +1,52 @@
+/*
```
I'd move this file one folder up: I don't think it's a mock

jori-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+find_package(Zephyr COMPONENTS unittest HINTS $ENV{ZEPHYR_BASE})
+include(${ZEPHYR_BASE}/tests/bluetooth/host/cmake.txt)
+include(${ZEPHYR_BASE}/tests/bluetooth/host/buf/cmake.txt)
+target_sources(testbinary PRIVATE ${host_module} ${module_mocks} ${host_mocks})
```
can you add some newlines for better readability ?
```cmake
project(bluetooth_bt_buf_get_cmd_complete)

find_package(Zephyr COMPONENTS unittest HINTS $ENV{ZEPHYR_BASE})

include(${ZEPHYR_BASE}/tests/bluetooth/host/cmake.txt)
include(${ZEPHYR_BASE}/tests/bluetooth/host/buf/cmake.txt)

target_sources(testbinary PRIVATE ${host_module} ${module_mocks} ${host_mocks})
```

jori-nordic reviewed (dismissed):
just 2 small nits and for me it's ready to be squashed and merged.

jori-nordic reviewed (dismissed):
Could you rename the commit message summary:

`tests: Bluetooth: host: Add unit tests for buf.c`

ahmedmoheb-nordic commented:
Hi @yperess and @thoh-ot , could you review again or mark your change request to be resolved?

@yperess regarding moving from ZTest mocks to FFF, a following PR will do the required changes.

Thanks

jori-nordic commented:
Since ahmen didn't mention it, here's the following PR he's talking about @yperess https://github.com/zephyrproject-rtos/zephyr/pull/48461

tejlmand commented on tests/bluetooth/host/buf/cmake.txt:
```diff
@@ -0,0 +1,22 @@
+#
```
we should use this principle.

It has been commented here:
https://github.com/zephyrproject-rtos/zephyr/pull/40073#issuecomment-1041451330

unfortunately I was not invited to that party.

Blocking this PR for time being and will provide a proper prosal after taking closer look at this code.

tejlmand commented on tests/bluetooth/host/cmake.txt:
```diff
@@ -0,0 +1,15 @@
+)
+
+add_definitions(-include kconfig.h)
+add_definitions(-include ztest.h)
```
why do we need this ?

Other tests seems to be fine by using `#include <ztest.h>`
https://github.com/zephyrproject-rtos/zephyr/blob/b88c8e13634437723e294b213876672f31a78e09/tests/kernel/smp/src/main.c#L8

Do you really want `ztest.h` included in ALL source files ?

ahmedmoheb-nordic replied:
I tried to remove it but compilation process didn't go well as some files that aren't direct project sources require it.
I think this file is required as there is no `autoconf.h`

tejlmand reviewed (changes requested):
Some extra comments.
I'll take a look at the unittest / Kconfig support.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/testcase.yaml:
```diff
@@ -0,0 +1,14 @@
+     type: unit
+  bluetooth.host.bt_buf_get_cmd_complete.hci_acl_flow_control:
+     type: unit
+     extra_args: COMPILE_SWITCH_PARAM=CONFIG_BT_HCI_ACL_FLOW_CONTROL
```
I already tried your suggested way and seems not be working (at least when the test type is set to unit testing using `type: unit`)

tejlmand replied:
> I already tried your suggested way and it seems not be working

yes, because Kconfig is not supported, which is why I also wrote :wink: : 
> What is done in this PR is to invent a new config scheme instead of supporting Kconfig.
> 
> I will take a look to see if I can get Kconfig supported with unittest so that both this and the comments in https://github.com/zephyrproject-rtos/zephyr/pull/40073 can be cleaned up.

tejlmand replied:
just made the first comment so we remember what to clean up and not forgetting.
Little you can do atm.

ahmedmoheb-nordic commented on tests/bluetooth/host/buf/cmake.txt:
```diff
@@ -0,0 +1,22 @@
+  ${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks
+)
+
+FILE(GLOB host_module
```
How can I create list of source files without using `GLOB`?

tejlmand replied:
easy:
```
set(host_module
    ${ZEPHYR_BASE}/subsys/bluetooth/host/buf.c
)
```

tejlmand replied:
also note, that if the file `${ZEPHYR_BASE}/subsys/bluetooth/host/buf.c` doesn't exists then the `host_module` just becomes an empty list and you feed an empty list of source files to the library which will not give an error.

But if using the `set()` approach and the file doesn't exists, then CMake will throw an error.

tejlmand commented:
> I'll take a look at the unittest / Kconfig support.

Kconfig support is now ready here: #48732

@ahmedmoheb-nordic  please try it out for Kconfig settings together with this PR and provide feedback.

jori-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+find_package(Zephyr COMPONENTS unittest HINTS $ENV{ZEPHYR_BASE})
+
+include(${ZEPHYR_BASE}/tests/bluetooth/host/CMakeLists.txt)
```
It's my understanding that if we use `include`, we should have a `.cmake` file.
Did you try using libraries instead?

ahmedmoheb-nordic replied:
Where can I find such guidelines about file names and things to do and things not to do?

I just used one of the suggested ways to name the files by **Torsten**
`CMake files should be named either:
CMakeLists.txt or <propername>.cmake.`

Although I wasn't in favor of using the same file name `CMakeLists.txt ` for all files, but I tried to avoid the dilemma of selecting the proper name.

On the other hand, I tried to build libraries, but It didn't work because it depends on auto generated files while building.
I will try it again when the change made by **Torsten** to use `Kconfig `and `prj.conf `files is merged.

So, for now, do you recommend change the file names again?
If yes, do you think a proper file name can be the parent folder name? (e.g. `host_mocks`, `buf_mocks`, `keys_mocks` ...etc)

jori-nordic replied:
I'm just parroting what I see on the cmake docs :p 
[Here](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Getting%20Started.html#basic-cmake-usage) it describes that the projects (libraries and executables) will be defined in `CMakeLists.txt` files.
[And here (last snippet)](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Writing%20CMakeLists%20Files.html#variable-scope) we see that they use `lowercasename.cmake` for files that are supposed to be included.

jori-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+
+if (NOT BOARD STREQUAL unit_testing)
```
do we still have to do this after torstein's PR?

ahmedmoheb-nordic replied:
I don't know what made this check redundant, but it seems not to be needed any more

jori-nordic commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+find_package(Zephyr COMPONENTS unittest HINTS $ENV{ZEPHYR_BASE})
+
+include(${ZEPHYR_BASE}/tests/bluetooth/host/host_mocks.cmake)
```
I think you already tried using a library and had a bunch of issues. Could you try again with torstein's changes to see if that resolves the issues?

ahmedmoheb-nordic replied:
Yes, it's on my to-do list

jori-nordic replied:
yeah I think at this point, @tejlmand is the one that can help.

ahmedmoheb-nordic replied:
With some help from **Torsten**, I was able to bypass that issue and explicitly add the `autoconf.h` while building the library.
That is done using
`zephyr_library_compile_options(-imacros ${AUTOCONF_H})`

But currently, I'm encountering a different problem.
The build process is now terminated with `fatal error: kobj-types-enum.h: No such file or directory`.
This file is automatically generated under the build directory and I can find it.
I already added the expected path of that file using `zephyr_library_include_directories(${CMAKE_BINARY_DIR}/zephyr/include/generated)`.

I think that the file is generated at some point after trying to build the library or building the library requires that all files are there before starting the build process.

Any suggestions?

tejlmand replied:
please find the rough framework here:
https://github.com/ahmedmoheb-nordic/zephyr/pull/5

jori-nordic commented on tests/bluetooth/host/host_mocks.cmake:
```diff
@@ -0,0 +1,15 @@
+  ${ZEPHYR_BASE}/tests/bluetooth/host/host_mocks/assert.c
+)
+
+add_definitions(-include kconfig.h)
```
I thought we didn't need this anymore

ahmedmoheb-nordic replied:
Yes, I already removed it

ahmedmoheb-nordic commented:
> > I'll take a look at the unittest / Kconfig support.
> 
> Kconfig support is now ready here: #48732
> 
> @ahmedmoheb-nordic please try it out for Kconfig settings together with this PR and provide feedback.

@tejlmand I've rebased my branch on top of your modifications and modified the files, Could you re-check?

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add mocks for buf.c

Add required mocks to be able to compile and test /bluetooth/host/buf.c

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_buf_get_rx()

Unit test project for bt_buf_get_rx().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_buf_get_cmd_complete()

Unit test project for bt_buf_get_cmd_complete().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_buf_get_evt()

Unit test project for bt_buf_get_evt().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_buf_get_type()

Unit test project for bt_buf_get_type().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

tejlmand committed:
cmake: create test_interface interface library

Creating a test_interface allows more complex unittest projects to
create additional libraries, for example mock libs, that can be re-used
between test projects.

To support the possibility of such libs to build with same flags as the
testbinary the test_interface is now populated with the public compile
and include flags from testbinary.

testbinary has been updated to use test_interface.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented on subsys/bluetooth/host/buf.c:
```diff
@@ -141,3 +141,36 @@ struct net_buf *bt_buf_get_evt(uint8_t evt, bool discardable,
 		return bt_buf_get_rx(BT_BUF_EVT, timeout);
 	}
 }
+
```
nothing wrong on this line but cannot place comment on commit message.

Commit message says:
> host: Add/Copy mocks for buf.c

are we copying mocks from another location, or how is the `copy` supposed to be understood.
I don't hope we a duplicating code with this commit.

If you are copying files, could you please be a bit more specific and reference the origin of the file in the commit message.

ahmedmoheb-nordic replied:
No, there is no any duplication here, may be the message is misleading.
I will edit the commit message to make it clear

tejlmand commented on tests/bluetooth/host/buf/bt_buf_get_rx/testcase.yaml:
```diff
@@ -0,0 +1,20 @@
+     type: unit
+  bluetooth.host.bt_buf_get_rx.hci_acl_flow_control:
+     type: unit
+     # enable CONFIG_BT_HCI_ACL_FLOW_CONTROL
+     extra_configs:
+      - CONFIG_BT_CENTRAL=y
```
the test is called: `hci_acl_flow_control`

so in addition / instead of a comment, maybe make it explicit that this test is enabling this, although it's the default.
```suggestion
     extra_configs:
      - CONFIG_BT_CENTRAL=y
      - CONFIG_BT_HCI_ACL_FLOW_CONTROL=y
```

that will make it more clear to the reader what we are testing, and removes the need for a comment.

tejlmand commented on tests/bluetooth/host/buf/bt_buf_get_cmd_complete/testcase.yaml:
```diff
@@ -0,0 +1,20 @@
+     type: unit
+  bluetooth.host.bt_buf_get_cmd_complete.hci_acl_flow_control:
+     type: unit
+     # enable CONFIG_BT_HCI_ACL_FLOW_CONTROL
+     extra_configs:
+      - CONFIG_BT_CENTRAL=y
```
same comment regarding explicit setting of `CONFIG_BT_HCI_ACL_FLOW_CONTROL=y` to emphasize that is what we are testing.

tejlmand commented on tests/bluetooth/host/buf/bt_buf_get_evt/testcase.yaml:
```diff
@@ -0,0 +1,20 @@
+     type: unit
+     # enable CONFIG_BT_HCI_ACL_FLOW_CONTROL
+     extra_configs:
+      - CONFIG_BT_CENTRAL=y
```
same comment regarding explicit setting of `CONFIG_BT_HCI_ACL_FLOW_CONTROL=y` to emphasize that is what we are testing.

tejlmand reviewed (commented):
much better looking, just noticed couple of minor nits.

Only thing is that I really would like @cvinayak or @Thalley to take a look at this comment:
https://github.com/zephyrproject-rtos/zephyr/pull/47794#discussion_r958468842

so will give a little time for them to take a look, else will +1 the PR.

tejlmand commented:
> Only thing is that I really would like @cvinayak or @Thalley to take a look at this comment: [#47794 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/47794#discussion_r958468842)
> 
> so will give a little time for them to take a look, else will +1 the PR.

also summoning: @alwa-nordic Aleksande

ahmedmoheb-nordic commented:
> > Only thing is that I really would like @cvinayak or @Thalley to take a look at this comment: [#47794 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/47794#discussion_r958468842)
> > so will give a little time for them to take a look, else will +1 the PR.
> 
> also summoning: @alwa-nordic Aleksande

Also, @jori-nordic can assist, I already discussed that part with him

cvinayak commented on subsys/bluetooth/host/buf.c:
```diff
@@ -141,3 +141,34 @@ struct net_buf *bt_buf_get_evt(uint8_t evt, bool discardable,
 	}
 }
+
+#ifdef ZTEST_UNITTEST
```
The unit test specific code was added as part of the development of new LL Control Procedure and I believe these method/model was agreed by the team to access file static memory by the unit test code.

There are other legacy bare-metal unit test code that in fact run on target at controller initialization, but these run as in-system tests, ex `CONFIG_BT_CTLR_TEST`, but a legacy one is `MAYFLY_UT` that tests thread and ISR race condition, this something needs clean up as it appears as dead code now.

As long as the intention of the unit test code present in this PR only is (cleanly) used by unit tests, I have no objection.

tejlmand replied:
great. thanks for approving the principle.

tejlmand reviewed (dismissed):
lgtm, thanks for the effort 👍

jhedberg commented on subsys/bluetooth/host/buf.c:
```diff
@@ -141,3 +141,36 @@ struct net_buf *bt_buf_get_evt(uint8_t evt, bool discardable,
+	return &num_complete_pool;
+}
+#endif
+#endif
```
Could you make this `#endif /* ZTEST_UNITTEST */` to make it clear which condition it is terminating?  It’s in general a good idea to do that, especially for long blocks and when you have many ifdefs in the same file.

ahmedmoheb-nordic commented:
> lgtm, thanks for the effort +1

Thanks to vote again for the changes

tejlmand reviewed (approved):
lgtm
